[TechDocs] Optional static resource caching

iamEAP opened the pull request:
## Hey, I just made a Pull Request!

Introduces a caching layer for static assets served through the TechDocs backend.  We've found this necessary internally due to high latencies (600-1000ms per file) when reading TechDocs content from Google Cloud Storage, our preferred storage solution, especially in a multi-region deployment configuration.  Other adopters have reported [similar trouble](https://discord.com/channels/687207715902193673/714754240933003266/860139912324972544) with latency.

### Architecture

![Screenshot 2021-11-12 at 15 05 10](https://user-images.githubusercontent.com/3496491/141480136-9441e416-cd52-4a03-9864-46d4b24c8b37.png)

### Implementation:
- Built on top of the [cache manager in backend-common](https://github.com/backstage/backstage/pull/5551).
- Caching layer is implemented in `@backstage/plugin-techdocs-backend`, so that each storage-specific publisher implementation doesn't have to be aware of caching.
- Supports long-TTLs (for infrequently updated sites) by hooking into the docs synchronization mechanism:
  - If a difference is found between `techdocs_metadata.json`'s build timestamps (b/w cache and original storage), all objects are invalidated in the cache store and an `updated` message is sent to the front-end (popping the "a new version of these docs is available..." message / refresh button)
  - For those adopters who run local generation in combination with external storage, cache invalidation is handled out-of-the-box by the `DocsBuilder`.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

ottosichert commented on docs/features/techdocs/architecture.md:
```diff
@@ -58,12 +58,12 @@ Similar to how it is done in the Basic setup, the TechDocs Reader requests
-
-_Feel free to suggest better ideas to us in #docs-like-code channel in Discord
-or via a GitHub issue._
+Depending on your chosen cloud storage provider and its real-world proximity to
+your backend server, there may be a comparably high amount of latency when
+loading TechDocs sites using this deployment approach. If you encounter this,
+you can optionally configure the `techdocs-backend` to cache responses in a
+cache store
+[supported by Backstage](../../overview/architecture-overview.md#cache).
```
Have we considered the effect of deployment strategies of the Backstage instance? I.e. if running on multiple nodes (e.g. availability zones), we might want to give a hint on the tradeoff between storage and speed. So in order to maximize speed one might consider caching at the edge, or having one cache per region if storage is a concern

iamEAP replied:
Indeed, and this more closely approximates how we run things internally now.  Wanted to keep the blurb here high-ish level since it's about architecture generally and the potential configurations with benefits / tradeoffs get complicated quickly.  Good opportunity to expand on all of this in a "deployment configurations and recommendations" section of its own, though.

ottosichert commented on packages/techdocs-common/src/stages/publish/local.ts:
```diff
@@ -89,24 +89,29 @@ export class LocalPublish implements PublisherBase {
 
     return new Promise((resolve, reject) => {
-      fs.copy(directory, publishDir, err => {
+      fs.copy(directory, publishDir, async err => {
```
`fs-extra` does not tread the callback as a Promise explicitly, hence why it might be preferred to stay in Promise land even before making the call to `fs.copy`. This can be achieved by defining the `publish` function as `async` and simply `await`ing `fs.copy` as follows:

```js
async publish(foo): bar {
  // sync code
  try {
    await fs.copy(directory, publishDir);
  } catch (error) {
    this.logger.debug('baz');
    throw error;
  }
  
  const baseUrl = await this.discovery.getBaseUrl('techdocs');
  // async stuff
  return { remoteUrl: 'quux' };
}
```

iamEAP replied:
For the sake of PR legibility, I tried to keep the diff really small, but this does make things _a lot_ more readable (and you did the review already anyway!), so I'll refactor.

ottosichert commented on plugins/techdocs-backend/api-report.md:
```diff
@@ -9,12 +9,13 @@ import express from 'express';
 
 // @public (undocumented)
-export function createRouter({ preparers, generators, publisher, config, logger, discovery, }: RouterOptions): Promise<express.Router>;
+export function createRouter({ preparers, generators, publisher, config, logger, discovery, cache: cacheManager, }: RouterOptions): Promise<express.Router>;
```
I'm not 100% sure where `cacheManager` is defined, as we are importing `PluginCacheManager` above. Additionally, are we intending to provide a default value? This might be overridden by `undefined` in previous function signatures when we are always passing `cache` as value, even when it's optional

iamEAP replied:
Think this was simplified/clarified as a result of our OutOfTheBox/Recommended typing work a couple weeks back.

ottosichert commented on plugins/techdocs-backend/src/DocsBuilder/builder.ts:
```diff
@@ -192,11 +197,16 @@ export class DocsBuilder {
     });
 
+    // Invalidate the cache for any published objects.
+    if (this.cache && published && published?.objects?.length) {
```
_Nit_: When `grep`-ing through the source I can see we have roughly an equal amount of implicit non-zero length checking and explicit. Do we have a convention whether to use `if (length)` vs `if (length > 0)`?

ottosichert commented on plugins/techdocs-backend/src/cache/TechDocsCache.ts:
```diff
@@ -0,0 +1,87 @@
+    try {
+      // Promise.race ensures we don't hang the client for long if the cache is
+      // temporarily unreachable.
+      const response = (await Promise.race([
+        this.cache.get(path),
+        new Promise(cancelAfter => setTimeout(cancelAfter, 1000)),
+      ])) as string | undefined;
```
In the case of network congestion, do we have a way to cancel get operations? We might run into an issue where we are blocking all available connections due to hanging connections

iamEAP replied:
Not a feature of the underlying cache client currently, unfortunately.

ottosichert commented on plugins/techdocs-backend/src/cache/TechDocsCache.ts:
```diff
@@ -0,0 +1,87 @@
+    }
+  }
+
+  async set(path: string, data: Buffer): Promise<void> {
+    this.logger.debug(`Writing cache entry for ${path}`);
+    this.cache
+      .set(path, data.toString('base64'))
+      .catch(e => this.logger.error('write error', e));
+  }
+
+  async invalidate(path: string): Promise<void> {
+    return this.cache.delete(path);
+  }
```
Wouldn't we need to apply timeouts here as well?

iamEAP replied:
The intention behind the timeout on the `get` was to ensure an end-user is at least able to eventually load the documentation they were trying to read in case of issues (without really being made aware of the underlying problem).

Here, the end-user is a TechDocs _author_, and it's probably more important that they be made aware of the issue (by allowing the underlying cache client to try and potentially throw an error) than swallow it up for handle it ourselves.

ottosichert commented on plugins/techdocs-backend/src/cache/cacheMiddleware.ts:
```diff
@@ -0,0 +1,139 @@
+      return realWrite(data, encoding, callback);
+    };
+
+    // When a socket is closed, if there were no errors and the data written
+    // over the socket should be cached, cache it as a base64-encoded string!
+    socket.on('close', hadError => {
+      if (writeToCache && !hadError) {
+        cache.set(reqPath, Buffer.concat(chunks));
+      }
+    });
```
Just curious: is a client completely done processing the response and the cache write is happening "in the background"? https://nodejs.org/api/stream.html#stream_event_close indicates it would be the case, so what happens if an exception is thrown at this place?

iamEAP replied:
Would guess the app would crash.  However, `cache.set` is just a wrapper around `this.cache.set()` in the underlying cache client, which has a `.catch()` attached that just logs the error.  So we should be safe, here.

ottosichert reviewed (commented):
Love the work on performance! (insert rocket emoji here)

iamEAP commented on app-config.yaml:
```diff
@@ -105,7 +105,8 @@ techdocs:
     type: 'local' # Alternatives - 'googleGcs' or 'awsS3' or 'azureBlobStorage' or 'openStackSwift'. Read documentation for using alternatives.
-
+  cache:
+    ttl: 60000 # 1 minute cache for demonstration purposes. You may wish to set this higher in production.
```
Will update to 1 hour for consistency; this is only ever loaded during local development of Backstage, so no worry about it accidentally leaking into a production config.  In real life, you're right, much higher values are probably warranted, depending on the underlying cache store.

andrewthauer commented on plugins/techdocs-backend/src/cache/TechDocsCache.ts:
```diff
@@ -0,0 +1,87 @@
+      // temporarily unreachable.
+      const response = (await Promise.race([
+        this.cache.get(path),
+        new Promise(cancelAfter => setTimeout(cancelAfter, 1000)),
```
Can this timeout be configurable?

iamEAP replied:
Good idea.

andrewthauer reviewed (commented):
Does this mean that if you want to use caching and publish docs from CI, that your CI system will need access to the hosted Backstage instance? If so, this feature won't be easily usable for orgs that use a cloud CI solution and have Backstage hosted behind say a VPN, etc?

iamEAP commented:
> Does this mean that if you want to use caching and publish docs from CI, that your CI system will need access to the hosted Backstage instance?

Hmm... Perhaps we can support both hitting the TechDocs backend over http and the cache store directly, depending on CLI configs? Then the cache store would be available outside the VPN, but at least it would have an established auth mechanism.

I suppose this kind of access pattern could/would become a wider issue if/when externally driven catalog updates become a thing (e.g. triggering an entity refresh via VCS web-hook)... Wonder how that might be handled.

andrewthauer commented:
> Hmm... Perhaps we can support both hitting the TechDocs backend over http and the cache store directly, depending on CLI configs? Then the cache store would have to be available outside the VPN, but at least it would have an established auth mechanism.

Sorry, not completely following. How would this work? Isn't the cache endpoint part of the techdocs backend already?

> I suppose this kind of access pattern could/would become a wider issue if/when externally driven catalog updates become a thing (e.g. triggering an entity refresh via VCS web-hook)... Wonder how that might be handled.

Yes, this is something I've started thinking about more and more. There are at least a couple of scenarios where having CI and/or webhooks/etc occurring externally would be beneficial.

iamEAP commented:
> Sorry, not completely following. How would this work? Isn't the cache endpoint part of the techdocs backend already?

I'm thinking it could be possible to provide the `techdocs-cli` with enough info to instantiate its own cache client and perform the invalidation itself, rather than having to hit the TechDocs API.  It would have already have the keys it needs to invalidate; it just needs connection details, which could be provided via flag, environment vars, or other config.

andrewthauer commented:
> I'm thinking it could be possible to provide the `techdocs-cli` with enough info to instantiate its own cache client and perform the invalidation itself, rather than having to hit the TechDocs API.  It would have already have the keys it needs to invalidate; it just needs connection details, which could be provided via flag, environment vars, or other config.

So are you suggesting the techdocs-cli would invalidate the underlying cache service directly?
Does this not mean the caching service would need to be publicly accessible as well if the techdocs-cli is running in a cloud hosted CI service?

dhenneke commented:
> > I'm thinking it could be possible to provide the `techdocs-cli` with enough info to instantiate its own cache client and perform the invalidation itself, rather than having to hit the TechDocs API.  It would have already have the keys it needs to invalidate; it just needs connection details, which could be provided via flag, environment vars, or other config.
> 
> So are you suggesting the techdocs-cli would invalidate the underlying cache service directly? Does this not mean the caching service would need to be publicly accessible as well if the techdocs-cli is running in a cloud hosted CI service?

How about using the `/sync` endpoint for this instead? It is currently a [no-op for the `external`](https://github.com/backstage/backstage/blob/976db8d9f867136abbc711d85fb4a5292b43b1de/plugins/techdocs-backend/src/service/router.ts#L174-L180) setting but it could instead e.g. fetch the `techdocs_metadata.json` from the backing storage and compare it to the cached one. If they differ, it could invalidate the cache for the entity. In case of an invalidation, it could also display the "you need to refresh the page to see the latest version" alert.

iamEAP commented:
> How about using the `/sync` endpoint for this instead? It is currently a [no-op for the `external`](https://github.com/backstage/backstage/blob/976db8d9f867136abbc711d85fb4a5292b43b1de/plugins/techdocs-backend/src/service/router.ts#L174-L180) setting but it could instead e.g. fetch the `techdocs_metadata.json` from the backing storage and compare it to the cached one. If they differ, it could invalidate the cache for the entity. In case of an invalidation, it could also display the "you need to refresh the page to see the latest version" alert.

This is intriguing!  I guess the trouble with this would be that we don't have the full context of which individual objects to invalidate.  Cache stores with wildcard delete support can get by, but `memcache` would not, for example.

Perhaps if we published like a `cache_manifest.json` file alongside the `techdocs_metadata.json` file containing currently published objects and compared those?

andrewthauer commented:
Wouldn't publishing the docs site in most cases invalidate the entire site rather then specific files? Wondering if there could be a separate TTL for checking a meta file timestamp. If that changed, the entire site is invalidated.

iamEAP commented:
> Wouldn't publishing the docs site in most cases invalidate the entire site rather then specific files? 

Yep.  But the list of files is only known at publish-time, and cache stores are not guaranteed to support something like `cache.invalidate('namespace/kind/component/*');`.  Memcache in particular (which is only supported cache store currently other than in-memory) does not.  I like the idea though, so I made the following changes.

### Updates
- Updated `techdocs-common` so that the list of files published to the bucket is written to `techdocs_metadata.json` at publish-time.
- Added a `doCacheSync` method to the DocsSynchronizer that's invoked when an external builder is configured and cache is enabled.  This compares build timestamps from the "source" file (loaded through the publisher directly) and the cached file (loaded by hair-pinning a request back through the service) and invalidates all related objects.
- Removed the cache invalidation endpoint.

### Disadvantages to this approach
- There is a possibility, depending on the underlying cache store used, that the cached `techdocs_metadata.json` is evicted prior to other cached assets for a site.  In this case, if the published docs were actually updated, the next build time comparison would result in a false match, and stale content would continue to be shown until other cached assets expired or were evicted.
- There is a possibility that the combined list of files used to invalidate caches is incomplete due to intermediary publishes that occur between a first sync check and a subsequent sync check.  This could result in, for example, files persisting in cache after they've been removed.  (Really only a problem once #6488 is in, and kind of a minor problem, at that)
- This approach isn't perfect in case when the TechDocs backend is horizontally scaled because the mechanism for preventing excessive metadata checks relies on a value stored in memory.

### Advantages to this approach
- Makes the cache architecture possible/secure in the case when Backstage is deployed behind a firewall but docs builds happen in a public/cloud CI provider.
- Avoids complicating the `techdocs-cli` publish command with additional arguments for knowing how to reach the TechDocs backend (also completely avoids having to deal with scenarios where the TechDocs backend is deployed in multiple, independent regions).
- You get a nice `There's a newer version! Click here to see it` message when there _is_ a newer version.

iamEAP commented:
Still a couple of todos too:

- [ ] Update the Architecture Diagram to reflect the new approach
- [ ] Make the cache get timeout configurable

iamEAP commented:
Rebased (and fixed bugs -- a lot has happened since I last touched this), made the cache read timeout configurable, and updated the architecture diagram / docs to reflect the new approach recommended by Andrew.

freben commented on packages/techdocs-common/src/stages/publish/local.ts:
```diff
@@ -112,27 +115,26 @@ export class LocalPublish implements PublisherBase {
     });
+
+    return {
+      remoteUrl: `${techdocsApiUrl}/static/docs/${entity.metadata.name}`,
```
`encodeURIComponent` on the entity parts of URLs just to be sure

freben commented on plugins/techdocs-backend/src/cache/TechDocsCache.ts:
```diff
@@ -0,0 +1,109 @@
+import { Config } from '@backstage/config';
+import { Logger } from 'winston';
+
+export class CacheInvalidationError extends Error {
```
[CustomErrorBase](https://github.com/backstage/backstage/blob/60020e8ae4a4b9814ec173042e1ea504b3c72d9e/packages/errors/src/errors/CustomErrorBase.ts#L35) from the errors package is a good helper for stuff like this.

freben commented on plugins/techdocs-backend/src/cache/cacheMiddleware.ts:
```diff
@@ -0,0 +1,94 @@
+import { Router } from 'express';
+import router from 'express-promise-router';
+import { Logger } from 'winston';
+import { TechDocsCache } from '.';
```
avoid dot imports

freben commented on plugins/techdocs-backend/src/cache/cacheMiddleware.ts:
```diff
@@ -0,0 +1,94 @@
+    socket.on('close', hadError => {
+      const content = Buffer.concat(chunks);
+      const head = content.toString('utf8', 0, 12);
+      if (writeToCache && !hadError && head === 'HTTP/1.1 200') {
```
well i mean HTTP/1.1 is pretty specific :)

freben reviewed (approved):
This is still wanted, right? I gave it a quick look and an unblocking appove from my end so you can get the others to give it a final look too.

ottosichert commented on plugins/techdocs-backend/src/cache/TechDocsCache.ts:
```diff
@@ -0,0 +1,105 @@
+        new Promise(cancelAfter => setTimeout(cancelAfter, this.readTimeout)),
+      ])) as string | undefined;
+
+      if (response !== undefined) {
```
By conflating the origin of `undefined` coming from either the `setTimeout` or an empty `cache.get`, I think we might miss catching errors in the underlying cache. It would be worthwhile to have the option to monitor if 100% of cache misses are caused by timeouts or return immediately and hint to errors with the cache itself

iamEAP replied:
I'd love to get this level of detail in some kind of observability framework, since even as is, it's only `debug`-level log statements, which probably won't be captured anywhere in production.   For now, you can get a sense of when things go awry when response times start to be suspiciously correlated to the `readTimeout` configuration value (I've seen blips of this in our internal deployment of this patch-set, for example).

ottosichert commented on plugins/techdocs-backend/src/cache/cacheMiddleware.ts:
```diff
@@ -0,0 +1,94 @@
+  // loaded from cache. Cache key is the object's path (after `/static/docs/`).
+  cacheMiddleware.use(async (req, res, next) => {
+    const socket = res.socket;
+    const isCacheable = req.path.includes('/static/docs/');
```
Would it be possible to use `startsWith`? Should be unlikely but prevents from unrelated resources to be marked cacheable

iamEAP replied:
I had assumed that `req.path` would include the full `path`-part of the request URL, but it turns out that in the context of express middleware, the mount point is removed; so `startsWith` is indeed the safer option!  Will update.

From express docs: 

> _When called from a middleware, the mount point is not included in req.path. See app.use() for more details._

iamEAP replied:
^ It appears supertest doesn't actually behave this way, but the real app does, at least.  So I tweaked the tests to be conformant.

ottosichert commented on plugins/techdocs-backend/src/cache/cacheMiddleware.ts:
```diff
@@ -0,0 +1,94 @@
+      const content = Buffer.concat(chunks);
+      const head = content.toString('utf8', 0, 12);
+      if (writeToCache && !hadError && head.match(/HTTP\/\d\.\d 200/)) {
+        cache.set(reqPath, content);
```
To indicate this call is async, we might make the whole function `async` and `await` here, as the Promise is eagerly evaluated but increases clarity of (a)synchronicity

iamEAP replied:
This is great feedback.

ottosichert reviewed (approved):
This is great work, and I think adopters and users will largely benefit from this implementation, thanks!

There are minor thoughts I added to the review, however I do think we should merge this as soon as possible :)

iamEAP commented on plugins/techdocs-backend/src/cache/TechDocsCache.ts:
```diff
@@ -0,0 +1,105 @@
+    config: Config,
+    { cache, logger }: { cache: CacheClient; logger: Logger },
+  ) {
+    const readTimeout =
+      config.getOptionalNumber('techdocs.cache.readTimeout') || 1000;
```
Good thought on the `0` value!  Will update.

As for the default handling in the constructor: the pattern I've tended to follow with these `fromConfig` static methods (especially when the constructor is private) is to keep the constructor dumb and most logic in the static method?  But 🤷‍♂️

iamEAP commented:
Thanks @ottosichert!  Love the feedback.

The above should take that into account + fix a lingering lint error that came up as a result of rebase.  If all's green, feel free to merge!  ...Probably better that y'all boop the button instead of me while I'm OOO!

ottosichert commented:
Awesome!! Let's get this merged then, thanks again for your efforts on this @iamEAP 💪
